agent/pkg/proxy/socks5_proxy.go

```diff
@@ -0,0 +1,285 @@
+package proxy
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	"github.com/kubeedge/edgemesh/agent/pkg/proxy/protocol"
+	"github.com/kubeedge/edgemesh/agent/pkg/tunnel"
+	"github.com/kubeedge/edgemesh/common/constants"
+)
+
+const (
+	LabelKubeedge string = "kubeedge=edgemesh-agent"
+	AgentPodName  string = "edgemesh-agent"
+
+	// Version is socks5 version
+	Version byte = 0x05
+
+	// DefaultMethod is No certification required
+	DefaultMethod byte = 0x00
+
+	Success byte = 0x00
+
+	// ATYPIPv4 is ipv4 address type
+	ATYPIPv4 byte = 0x01 // 4 octets
+	// ATYPDomain is domain address type
+	ATYPDomain byte = 0x03 // The first octet of the address field contains the number of octets of name that follow, there is no terminating NUL octet.
+	// ATYPIPv6 is ipv6 address type
+	ATYPIPv6 byte = 0x04 // 16 octets
+
+	// CmdConnect is connect command
+	CmdConnect byte = 0x01
+)
+
+// DefaultResponse is Socks5 returns data by default
+var DefaultResponse = []byte{Version, Success, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
+
+type Request struct {
+	Version     byte
+	Command     byte
+	Rsv         byte
+	AddressType byte
+	DstAddr     string
+	DstPort     int32
+}
+
+type SocksHandle struct {
+	Request *Request
+}
+
+type Socks5Proxy struct {
+	TCPProxy    *protocol.TCPProxy
+	kubeClient  kubernetes.Interface
+	NodeName    string
+	SocksHandle *SocksHandle
+}
+
+func (s *SocksHandle) ParsingConnect(conn net.Conn) (err error) {
+	err = s.handShake(conn)
+	if err != nil {
+		return err
+	}
+
+	err = s.NewRequest(conn)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SocksHandle) handShake(conn net.Conn) (err error) {
+	data := make([]byte, 2)
+	_, err = conn.Read(data)
+	if err != nil {
+		return err
+	}
+
+	if data[0] != Version {
+		return fmt.Errorf("invalid version")
+	}
+
+	if data[1] == 0 {
+		return fmt.Errorf("method length error")
+	}
+
+	ms := make([]byte, int(data[1]))
+	_, err = conn.Read(ms)
+	if err != nil {
+		return err
+	}
+
+	flag := false
+	var m byte
+	for _, m = range ms {
+		if m == DefaultMethod {
+			flag = true
+		}
+	}
+	if flag == false {
+		return fmt.Errorf("this method is not yet supported")
+	}
+
+	_, err = conn.Write([]byte{Version, Success})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SocksHandle) NewRequest(conn net.Conn) (err error) {
```

Comment:
If our method is copied from other github project and update for our own project, we should add a link 
```suggestion
# copy from https://github.com/txthinking/socks5/blob/e03c1217a50bd1363a2aaf58290da622256704fa/server_side.go#L125 and update
func (s *SocksHandle) NewRequest(conn net.Conn) (err error) {
```